.github/workflows/e2e-tests.yml

```diff
@@ -12,20 +12,87 @@ on:
         type: string
 
 jobs:
+
   e2e:
-    name: "E2E Tests"
+    name: E2E Tests - Shard ${{ matrix.shard }} of ${{ matrix.total_shards }}
+    runs-on: ubuntu-latest
+
+    strategy:
+        matrix:
+          shard: [1, 2, 3]
+          total_shards: [3]
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: lts/*
+
+      - name: Run e2e tests (Shard ${{ matrix.shard }}/${{ matrix.total_shards }})
+        run: |
+          make e2e-test \
+            APP_NAME=${{ inputs.app_name }} \
+            BASE_URL=${{ inputs.service_endpoint }} \
+            CURRENT_SHARD=${{ matrix.shard }} \
+            TOTAL_SHARDS=${{ matrix.total_shards }}
+
+      - name: Verify blob report directory after tests
+        run: |
+          echo "Contents of blob-report directory on host:"
+          ls -R ./e2e/blob-report || echo "blob-report directory not found"
+
+      - name: Upload Blob Report
+        if: ${{ !cancelled() }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: all-blob-reports-shard-${{ matrix.shard }}
+          path: ./e2e/blob-report
+          retention-days: 1
+
+  create-report:
+    name: Create Merged Test Report
+    if: ${{ !cancelled() }}
+    needs: [e2e]
     runs-on: ubuntu-latest
 
     steps:
       - name: Checkout code
         uses: actions/checkout@v4
 
-      # Run the Playwright tests inside the Docker container
-      - name: Run E2E tests in Docker container
-        run: make e2e-test APP_NAME=${{ inputs.app_name }} BASE_URL=${{ inputs.service_endpoint }}
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+            node-version: lts/*
+
+      - name: Install dependencies in ./e2e
+        run: make e2e-setup-ci
```

Comment:
The `create-report` job has to be in a separate "job" that runs after all the shard runs - as in it can't really be done in any one of the shard jobs - (unless you can reach out across job runs and merge reports? I doubt this is possible). Therefore, for this separate job would have to spin up a new container.

I'm not sure there's any need to increase our CI job run time to spin up a container to do a merge. And merging reports probably only needs to run in CI. 

Locally I don't think we'd ever want to shard - so the report there is just html in the `playwright-report` folder.

Reply:
It also raises the question - should `e2e` job (running of tests)  even be run in a container in CI? It would speed things up if we didn't have to use docker in CI.

 I think we should still use docker in CI - the CI run should be consistent to how you run it locally in the container. E2E tests can be flaky and I think it makes sense to eliminate any environment issues at play both locally and in the CI run.

Reply:
That makes sense thanks for explaining. I still think it'd be valuable to convert the line `run: npx playwright merge-reports --reporter html ./all-blob-reports` to a make target like `run: make e2e-merge-report` or something like that so that we can test the sharding locally. I think also that in order to test sharding locally, when we download the blob artifacts we should download them to the same place that they were in originally.

so in other words locally we can do
```
make e2e-test APP_NAME=app BASE_URL=http://host.docker.internal:3001 TOTAL_SHARDS=3 CURRENT_SHARD=1
make e2e-test APP_NAME=app BASE_URL=http://host.docker.internal:3001 TOTAL_SHARDS=3 CURRENT_SHARD=2
make e2e-test APP_NAME=app BASE_URL=http://host.docker.internal:3001 TOTAL_SHARDS=3 CURRENT_SHARD=3
make e2e-merge-reports
```

and in CI we'd do something like

```
jobs:
  e2e:
    matrix:
      shard: [1,2,3]
      total_shards: [3]
    steps:
      - run: make e2e-test APP_NAME=app BASE_URL=http://host.docker.internal:3001 TOTAL_SHARDS=${ matrix.total_shards } CURRENT_SHARD=${ matrix.shard }
      - # upload artifact from ./e2e/blob-report/*
  create-report:
    steps:
      - # download artifact to ./e2e/blob-report/
      - run: make e2e-merge-reports
```

Reply:
I've opened the PR on template infra to merge part 1